var/spack/repos/builtin/packages/py-h5py/package.py

```diff
@@ -42,26 +43,28 @@ class PyH5py(PythonPackage):
     depends_on("python@3.7:", type=("build", "run"), when="@3.2:")
 
     # Build dependencies
-    depends_on("py-cython@0.23:", type="build", when="@:2")
-    depends_on("py-cython@0.29:", type=("build"), when="@3: ^python@:3.7")
-    depends_on("py-cython@0.29.14:", type=("build"), when="@3: ^python@3.8.0:3.8")
-    depends_on("py-cython@0.29.15:", type=("build"), when="@3: ^python@3.9.0:")
+    depends_on("py-cython@0.23:0", type="build", when="@:2")
+    depends_on("py-cython@0.29:0", type=("build"), when="@3:")
+    depends_on("py-cython@0.29.14:0", type=("build"), when="@3:3.7 ^python@3.8.0:3.8")
+    depends_on("py-cython@0.29.15:0", type=("build"), when="@3:3.7 ^python@3.9.0:")
     depends_on("py-pkgconfig", type="build")
     depends_on("py-setuptools", type="build")
+    depends_on("py-setuptools@61:", type="build", when="@3.8.0:")
     depends_on("py-wheel", type="build", when="@3:")
 
     # Build and runtime dependencies
     depends_on("py-cached-property@1.5:", type=("build", "run"), when="^python@:3.7")
     depends_on("py-numpy@1.7:", type=("build", "run"), when="@:2")
-    depends_on("py-numpy@1.14.5:", type=("build", "run"), when="@3: ^python@3.7.0:3.7")
-    depends_on("py-numpy@1.17.5:", type=("build", "run"), when="@3: ^python@3.8.0:3.8")
-    depends_on("py-numpy@1.19.3:", type=("build", "run"), when="@3: ^python@3.9.0:")
+    depends_on("py-numpy@1.14.5:", type=("build", "run"), when="@3:")
+    depends_on("py-numpy@1.17.5:", type=("build", "run"), when="@3:3.5 ^python@3.8.0:3.8")
+    depends_on("py-numpy@1.19.3:", type=("build", "run"), when="@3:3.5 ^python@3.9.0:")
     depends_on("py-six", type=("build", "run"), when="@:2")
 
     # Link dependencies (py-h5py v2 cannot build against HDF5 1.12 regardless
     # of API setting)
     depends_on("hdf5@1.8.4:1.11 +hl", when="@:2")
-    depends_on("hdf5@1.8.4:1.12 +hl", when="@3:")
+    depends_on("hdf5@1.8.4:1.12 +hl", when="@3:3.7")
+    depends_on("hdf5@1.8.4:1.14 +hl", when="@3.8:")
 
     # MPI dependencies
     depends_on("hdf5+mpi", when="+mpi")
```

Comment:
mpi4py dep needs to be updated too:
```python
    RUN_REQUIRES.append('mpi4py >=3.0.2')                                                
    SETUP_REQUIRES.append("mpi4py ==3.0.2; python_version<'3.8'")                        
    SETUP_REQUIRES.append("mpi4py ==3.0.3; python_version=='3.8.*'")                     
    SETUP_REQUIRES.append("mpi4py ==3.1.0; python_version=='3.9.*' or python_version=='3.10.*'")
    SETUP_REQUIRES.append("mpi4py ==3.1.4; python_version>='3.11'")    
```
Honestly, this stuff is kinda annoying to track. Not sure why things need to be pinned to a specific version for each Python release. Should we just ignore this and say `mpi4py@3.0.2:` for all versions? Curious what @bryanherman @takluyver think.